apollo-ios-pagination/Sources/ApolloPagination/GraphQLQueryPager.swift

```diff
@@ -1,342 +1,207 @@
-        }
-      }
-      await activeTask?.value
-    }
-
-    public func subscribe(onUpdate: @MainActor @escaping (Result<Output, Error>) -> Void) -> AnyCancellable {
-      $currentValue.compactMap({ $0 }).sink { result in
-        Task {
-          await onUpdate(result)
-        }
-      }
-    }
-
-    /// Reloads all data, starting at the first query, resetting pagination state.
-    /// - Parameter cachePolicy: Preferred cache policy for first-page fetches. Defaults to `returnCacheDataAndFetch`
-    public func refetch(cachePolicy: CachePolicy = .fetchIgnoringCacheData) {
-      assert(firstPageWatcher != nil, "To create consistent product behaviors, calling `fetch` before calling `refetch` will use cached data while still refreshing.")
-      cancel()
-      fetch(cachePolicy: cachePolicy)
-    }
-
-    public func fetch() {
-      cancel()
-      fetch(cachePolicy: .returnCacheDataAndFetch)
-    }
-
-    /// Cancel any in progress fetching operations and unsubscribe from the store.
-    public func cancel() {
-      nextPageWatchers.forEach { $0.cancel() }
-      nextPageWatchers = []
-      firstPageWatcher?.cancel()
-      firstPageWatcher = nil
-
-      varMap = [:]
-      initialPageResult = nil
-      activeTask?.cancel()
-      activeTask = nil
-      subscribers.forEach { $0.cancel() }
-      subscribers.removeAll()
-    }
+private actor Subscriptions {
+  var subscriptions: Set<AnyCancellable> = []
 
-    /// Whether or not we can load more information based on the current page.
-    public var canLoadNext: Bool {
-      currentPageInfo?.canLoadMore ?? false
-    }
+  func store(subscription: AnyCancellable) {
+    subscriptions.insert(subscription)
+  }
+}
 
-    // MARK: - Private
+private class Completion {
+  var completion: ((Error?) -> Void)?
 
-    private func fetch(cachePolicy: CachePolicy = .returnCacheDataAndFetch) {
-      if firstPageWatcher == nil {
-        firstPageWatcher = GraphQLQueryWatcher(
-          client: client,
-          query: initialQuery,
-          resultHandler: { [weak self] result in
-            guard let self else { return }
-            Task {
-              await self.onInitialFetch(result: result)
-            }
-          }
-        )
-      }
-      firstPageWatcher?.refetch(cachePolicy: cachePolicy)
-    }
+  init(completion: ((Error?) -> Void)?) {
+    self.completion = completion
+  }
 
-    private func onInitialFetch(result: Result<GraphQLResult<InitialQuery.Data>, Error>) {
-      switch result {
-      case .success(let data):
-        initialPageResult = data.data
-        guard let firstPageData = data.data else { return }
-        if let latest {
-          let (_, nextPage) = latest
-          currentValue = .success((firstPageData, nextPage, data.source == .cache ? .cache : .fetch))
-        }
-      case .failure(let error):
-        currentValue = .failure(error)
-      }
+  func execute(error: Error?) async {
+    await MainActor.run { [weak self] in
```

Comment:
If we feel the need to run these on the `MainActor`, instead of `MainActor.run` internally here, could we mark the `completion` parameters on the `loadNext()` and `loadAll()` with `@MainActor`?

Maybe I'm not 100% clear on how `MainActor` works, but I'm thinking that way it is explicitly shown to the consumer that the completion block they pass in to those methods _will_ be run on the main thread. Whereas now, it's a hidden implementation detail.